posix/posix.c

```diff
@@ -671,6 +672,72 @@ int32_t red_transact(
 
     return PosixReturn(ret);
 }
+
+/** @brief Rollback to the previous committed state
+
+    Reliance Edge is a transactional file system.  All modifications, of both
+    metadata and filedata, are initially working state.  A transaction point
+    is a process whereby the working state atomically becomes the committed
+    state, replacing the previous committed state.  Whenever Reliance Edge is
+    mounted, including after power loss, the state of the file system after
+    mount is the most recent committed state.  Nothing from the committed
+    state is ever missing, and nothing from the working state is ever included.
+    This call cancel all modifications of the working state and get back to
+    the last commited state.
+
+    @param pszVolume    A path prefix identifying the volume to rollback.
+
+    @return On success, zero is returned.  On error, -1 is returned and
+            #red_errno is set appropriately.
+
+    <b>Errno values</b>
+    - #RED_EINVAL: Volume is not mounted; or @p pszVolume is `NULL`.
+    - #RED_EBUSY:  A discarded block is still referenced.
+    - #RED_ENOENT: @p pszVolume is not a valid volume path prefix.
+    - #RED_EUSERS: Cannot become a file system user: too many users.
+    - #RED_EBUSY:  There are still open handles for this file system volume.
+    - #RED_EROFS:  The file system volume is read-only.
+*/
+int32_t red_rollback(
+    const char *pszVolume)
+{
+    REDSTATUS   ret;
+
+    ret = PosixEnter();
+    if(ret == 0)
+    {
+        uint8_t bVolNum;
+
+        ret = RedPathVolumeLookup(pszVolume, &bVolNum);
+
+        if(ret == 0)
+        {
+            uint16_t    uHandleIdx;
+
+            /*  Do not rollback the volume if it still has open handles.
+            */
+            for(uHandleIdx = 0U; uHandleIdx < REDCONF_HANDLE_COUNT; uHandleIdx++)
+            {
+                const REDHANDLE *pHandle = &gaHandle[uHandleIdx];
+
+                if((pHandle->ulInode != INODE_INVALID) && (pHandle->bVolNum == bVolNum))
+                {
+                    ret = -RED_EBUSY;
+                    break;
+                }
+            }
+        }
+
+        if(ret == 0)
+        {
+            ret = RedCoreVolRollback();
+        }
+
```

Comment:
Need code here to deal with current working directories (CWDs) on this volume.  When we revert to the committed state, the CWDs may cease to exist.  One solution would be to reset the CWDs like `red_umount()` does.

Reply:
Should we reset it unconditionally or only if the corresponding inode has disappeared?

Reply:
It has to be reset unconditionally, for the reason you state: "Or could the inode referenced by CWD on the branch be a completely different inode in the original state."  In other words, the inode number may exist before and after the rollback, but there's no guarantee that it's the same inode.

Reply:
OK, thanks